spring-integration-test/src/main/java/org/springframework/integration/test/context/MockIntegrationContext.java

```diff
@@ -112,15 +120,31 @@ public void instead(String pollingAdapterId, MessageSource<?> mockMessageSource,
 		instead(pollingAdapterId, mockMessageSource, SourcePollingChannelAdapter.class, "source", autoStartup);
 	}
 
-	private void instead(String endpointId, Object mock, Class<?> endpointClass, String property,
+	public void instead(String consumerEndpointId, MessageHandler mockMessageHandler) {
+		instead(consumerEndpointId, mockMessageHandler, true);
+	}
+
+	public void instead(String consumerEndpointId, MessageHandler mockMessageHandler, boolean autoStartup) {
+		instead(consumerEndpointId, mockMessageHandler, IntegrationConsumer.class, "handler", autoStartup);
+		if (mockMessageHandler instanceof MessageProducer) {
+			Object targetMessageHandler = this.beans.get(consumerEndpointId);
+			if (targetMessageHandler instanceof MessageProducer) {
+				MessageChannel outputChannel = TestUtils.getPropertyValue(targetMessageHandler, "outputChannel",
+						MessageChannel.class);
+				((MessageProducer) mockMessageHandler).setOutputChannel(outputChannel);
+			}
```

Comment:
That would be quite surprising for the user to find his non-`MessageProducer` wrapped into `MH` :smile: Plain warning looks more safe.